### PR TITLE
avoid module-level initLogging for non-interactive CLI commands

### LINKED ISSUE #589: ocrd process does not work with ocropy in docker
When calling `cis-ocropy-binarize` tool with `ocrd process`

```
docker run --rm -u $(id -u) -v data/:/data -w /data -- ocrd/all:maximum-git ocrd process 'cis-ocropy-binarize -I OCR-D-IMG -O OCR-D-SEG-PAGE'
```

I get following error:

```
Traceback (most recent call last):
  File "/usr/bin/ocrd", line 33, in <module>
    sys.exit(load_entry_point('ocrd', 'console_scripts', 'ocrd')())
  File "/usr/lib/python3.6/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/usr/lib/python3.6/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/usr/lib/python3.6/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/lib/python3.6/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/lib/python3.6/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/build/core/ocrd/ocrd/cli/process.py", line 27, in process_cli
    run_tasks(mets, log_level, page_id, tasks, overwrite)
  File "/build/core/ocrd/ocrd/task_sequence.py", line 130, in run_tasks
    validate_tasks(tasks, workspace, page_id, overwrite)
  File "/build/core/ocrd/ocrd/task_sequence.py", line 99, in validate_tasks
    first_task.validate()
  File "/build/core/ocrd/ocrd/task_sequence.py", line 73, in validate
    param_validator = ParameterValidator(self.ocrd_tool_json)
  File "/build/core/ocrd/ocrd/task_sequence.py", line 54, in ocrd_tool_json
    self._ocrd_tool_json = json.loads(result.stdout)
  File "/usr/lib/python3.6/json/__init__.py", line 354, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python3.6/json/decoder.py", line 342, in decode
    raise JSONDecodeError("Extra data", s, end)
json.decoder.JSONDecodeError: Extra data: line 1 column 5 (char 4)
```
This was at least true for `cis-ocropy-deskew`. Calling the processes directly within the docker image works fine.

The docker image is up-to-date.

I am not sure if this is related to #540

### PR DESCRIPTION
**Pull request by bertsky**
Fix #589 (instead of #591)

This seems to work. Not so sure about the placement of `initLogging()` at the module level of `ocrd.cli.*` though...

### CODE REVIEW

**Comment by bertsky** [comment]
BTW, a simple way to test this is setting `logger_root` to `level=DEBUG` in `ocrd_logging.conf` (in `/etc` if in Docker or `~` native) – without this PR, you see DLL debug messages for `shapely.geos` in `ocrd --help` or `ocrd --version` (or similar). I tried to identify all CLI contexts which must not normally produce logging output in the patch (but might have missed some).

**Review by kba — APPROVED** [review]
Looks good. I'm not sure how intuitive I find it that importing `ocrd_utils.logging` effectively *disables* logging or that we now must do `initLogging` before doing anything else but at least that is well-defined behavior.

File: ocrd_utils/ocrd_utils/logging.py [hunk 1]
```diff
@@ -57,7 +57,7 @@ def getLevelName(lvl):
     lvl = _ocrdLevel2pythonLevel.get(lvl, lvl)
     return logging.getLevelName(lvl)
 
-def setOverrideLogLevel(lvl, silent=False):
```
[the remark this answers is not in the document]
Comment (bertsky) [reply ocrd_utils/ocrd_utils/logging.py:60]: Sorry, I thought this would not be needed anymore. I'll revert that part.

**Comment by bertsky** [comment]
> I'm not sure how intuitive I find it that importing `ocrd_utils.logging` effectively _disables_ logging or that we now must do `initLogging` before doing anything else but at least that is well-defined behavior.

Yes, it's not intuitive. The only other option I can imagine to get rid of logging in contexts like `--version`, `--help`, `--dump-json`, `ocrd-tool ...`, `bashlib ...` would be not changing ocrd_logging, but _not importing it at the module level_ anymore (but only when in a context where we really want it).

Also, we would still have to change logging setup/teardown for regression tests for this PR...

**Comment by bertsky** [comment]
> Yes, it's not intuitive. The only other option I can imagine to get rid of logging in contexts like `--version`, `--help`, `--dump-json`, `ocrd-tool ...`, `bashlib ...` would be not changing ocrd_logging, but _not importing it at the module level_ anymore (but only when in a context where we really want it).

But IIUC that would still not protect us against bad apple modules/libraries (like Matplotlib, Tensorflow...) which setup their logging in a way that they would interfere even without `basicConfig` from our side.

File: ocrd_utils/ocrd_utils/logging.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (bertsky) [reply ocrd_utils/ocrd_utils/logging.py]: Done. (I have also added more comments.)

**Comment by bertsky** [comment]
> Also, we would still have to change logging setup/teardown for regression tests for this PR...

I must confess I am perplexed about the 2 CI failures.
> `08:56:50.129 INFO ocrd.process.profile - [232 chars]\'\n`

Where could this possibly be originating from? `ocrd.processor.helpers.run_processor` only uses that logger once, with a completely different message content (`'Executing processor'...`)

**Comment by kba** [comment]
> I must confess I am perplexed about the 2 CI failures.
> 
> > `08:56:50.129 INFO ocrd.process.profile - [232 chars]\'\n`
> 
> Where could this possibly be originating from? `ocrd.processor.helpers.run_processor` only uses that logger once, with a completely different message content (`'Executing processor'...`)

The log messages themselves make sense, you're quoting the pytest abbreviation, the full output is

```
{"baz": "two"}
11:12:41.779 INFO ocrd.process.profile - Executing processor 'ocrd-test' took 0.000027s [--input-file-grp='OCR-D-IMG' --output-file-grp='OUTPUT' --parameter='{"baz": "two"}']
11:12:41.779 INFO ocrd.workspace - Saving mets '/tmp/ocrd-test-tempdir-boa5wmy4/mets.xml'
```

Now that initLogging needs to be explicitly called, these log messages interfere with the output of the processor.

I cannot, however, reproduce this locally. Can you?

**Comment by bertsky** [comment]
> The log messages themselves make sense, you're quoting the pytest abbreviation, the full output is

Oh, where do you get these from? Even the [download link](https://circleci.com/api/v1.1/project/github/OCR-D/core/1908/output/103/0?file=true&allocation-id=5f55f58b46f831119cf20ba5-0-build%2FF6B90B4) does not show this...

> Now that initLogging needs to be explicitly called, these log messages interfere with the output of the processor.

That's what I don't understand. Why didn't we get output from `ocrd.process.profile` when we used to `initLogging` unconditionally? It should never have worked ...

This test simulates a true CLI call, included standard logging (however implemented). In our unit tests, we can either test for non-logging output (where we should _disable_ logging) or for log output (with specific loggers _enabled_).

> I cannot, however, reproduce this locally. Can you?

No, not locally. All checks succeed.

**Comment by bertsky** [comment]
It looks like this is related to the differences between unittest's and click's runner. I have rebased and added a change to test that hypothesis. Thumbs crossed for CI... (Yes, that still does not give us conscious control over what kind of output we want in each test, but let's first get green lights on.)

**Comment by bertsky** [comment]
That _did_ work! Do you remember the reason for dealing with two different runners? (I fail to see the pattern.) Should we revisit tests as to whether or not they should or should not contain log output next, or are you satisfied with the current state for now?

> > Yes, it's not intuitive. The only other option I can imagine to get rid of logging in contexts like `--version`, `--help`, `--dump-json`, `ocrd-tool ...`, `bashlib ...` would be not changing ocrd_logging, but _not importing it at the module level_ anymore (but only when in a context where we really want it).
> 
> But IIUC that would still not protect us against bad apple modules/libraries (like Matplotlib, Tensorflow...) which setup their logging in a way that they would interfere even without `basicConfig` from our side.

I am not so enthusiastic about this PR myself, but what else could we do?

**Comment by kba** [comment]
> I am not so enthusiastic about this PR myself, but what else could we do?

Ah, cannot win prizes for elegance every time :) It's a good solution.

> Do you remember the reason for dealing with two different runners?

Output management, c.f. tests/base.py. pytest captures STDERR/STDOUT and log messages (to show in case of failure). In my experience, the click CliRunner works for simple tasks but for everything that requires STDOUT/STDERR (especially: separate), it's easier to run `cli.main` and catch SystemExit.

**Comment by bertsky** [comment]
> Output management, c.f. tests/base.py. pytest captures STDERR/STDOUT and log messages (to show in case of failure). In my experience, the click CliRunner works for simple tasks but for everything that requires STDOUT/STDERR (especially: separate), it's easier to run `cli.main` and catch SystemExit.

Understood. Then the last change is indeed warranted. (And the only place where log messages are checked via CLI is AFAICS  `TestDecorators.test_loglevel_invalid` – using CliRunner – and `TestLogging.testProcessorProfiling` – using `ocrd.run_processor` directly but also custom stream handler.)

**Pull request merged**